File: js/filter-parser.js
// DZX Filter Editor - Browser Filter Parser

class BrowserFilterParser {
    constructor() {
        this.conditionsPattern = /(\w+)\s+(.+?)(?=\s+\w+\s+|$)/g;
        this.actionsPattern = /(\w+(?:\w+\s+)*?)\s+([^\n]+)/g;
        this.colorPattern = /(\d+)\s+(\d+)\s+(\d+)(?:\s+(\d+))?/;
    }
    
    parseFile(file) {
        return new Promise((resolve, reject) => {
            const reader = new FileReader();
            
            reader.onload = (e) => {
                try {
                    const content = e.target.result;
                    const parsedData = this.parseContent(content);
                    resolve(parsedData);
                } catch (error) {
                    reject(error);
                }
            };
            
            reader.onerror = () => {
                reject(new Error('Failed to read file'));
            };
            
            reader.readAsText(file, 'utf-8');
        });
    }
    
    parseContent(content) {
        const lines = content.split('\n');
        const rules = [];
        let currentRule = null;
        let currentComment = null;
        
        for (let lineNum = 0; lineNum < lines.length; lineNum++) {
            const line = lines[lineNum].trim();
            
            // Skip empty lines
            if (!line) continue;
            
            // Handle comments
            if (line.startsWith('#')) {
                currentComment = line.substring(1).trim();
                continue;
            }
            
            // Handle Show/Hide statements
            if (line.startsWith('Show') || line.startsWith('Hide')) {
                if (currentRule) {
                    rules.push(currentRule);
                }
                
                const showHide = line.split()[0];
                currentRule = {
                    id: `rule_${rules.length + 1}`,
                    show_hide: showHide,
                    conditions: [],
                    actions: [],
                    comment: currentComment,
                    enabled: true
                };
                currentComment = null;
                continue;
            }
            
            // Handle conditions and actions
            if (currentRule) {
                if (this.isCondition(line)) {
                    const condition = this.parseCondition(line);
                    if (condition) {
                        currentRule.conditions.push(condition);
                    }
                } else if (this.isAction(line)) {
                    const action = this.parseAction(line);
                    if (action) {
                        currentRule.actions.push(action);
                    }
                }
            }
        }
        
        // Add the last rule
        if (currentRule) {
            rules.push(currentRule);
        }
        
        return {
            rules: rules,
            metadata: {
                total_rules: rules.length,
                parsed_at: new Date().toISOString()
            }
        };
    }
    
    isCondition(line) {
        const conditionKeywords = [
            'Class', 'BaseType', 'Rarity', 'AreaLevel', 'ItemLevel',
            'Sockets', 'Quality', 'Width', 'Height', 'DropLevel'
        ];
        const firstWord = line.split()[0];
        return conditionKeywords.includes(firstWord);
    }
    
    isAction(line) {
        const actionKeywords = [
            'SetTextColor', 'SetBorderColor', 'SetBackgroundColor',
            'SetFontSize', 'PlayEffect', 'MinimapIcon', 'CustomAlertSound',
            'PlayAlertSound', 'DisableDropSound', 'Continue'
        ];
        const firstWord = line.split()[0];
        return actionKeywords.includes(firstWord);
    }
    
    parseCondition(line) {
        try {
            const parts = line.split();
            if (parts.length < 2) return null;
            
            const conditionType = parts[0];
            let values = [];
            let operator = '==';
            
            // Handle different operators
            if (parts.length >= 3) {
                if (['>=', '<=', '>', '<', '==', '!='].includes(parts[1])) {
                    operator = parts[1];
                    values = [parts[2]];
                } else {
                    // Multiple values without operator
                    values = parts.slice(1);
                }
            } else {
                values = [parts[1]];
            }
            
            return {
                type: conditionType,
                operator: operator,
                values: values
            };
        } catch (error) {
            console.error(`Error parsing condition '${line}':`, error);
            return null;
        }
    }
    
    parseAction(line) {
        try {
            const parts = line.split();
            if (parts.length < 2) return null;
            
            const actionType = parts[0];
            let values = [];
            
            // Handle different action types
            if (actionType.includes('Color')) {
                // Parse RGB color values
                const colorMatch = this.colorPattern.exec(line);
                if (colorMatch) {
                    const r = parseInt(colorMatch[1]);
                    const g = parseInt(colorMatch[2]);
                    const b = parseInt(colorMatch[3]);
                    values = [r, g, b];
                    if (colorMatch[4]) {
                        values.push(parseInt(colorMatch[4]));
                    }
                } else {
                    values = parts.slice(1);
                }
            } else if (actionType === 'SetFontSize') {
                values = [parseInt(parts[1])];
            } else if (actionType === 'CustomAlertSound') {
                // Extract sound file and volume
                const soundMatch = line.match(/"([^"]+)"/);
                const volumeMatch = line.match(/(\d+)$/);
                if (soundMatch && volumeMatch) {
                    values = [soundMatch[1], parseInt(volumeMatch[1])];
                } else {
                    values = parts.slice(1);
                }
            } else {
                values = parts.slice(1);
            }
            
            return {
                type: actionType,
                values: values
            };
        } catch (error) {
            console.error(`Error parsing action '${line}':`, error);
            return null;
        }
    }
    
    validateSyntax(content) {
        const errors = [];
        const warnings = [];
        
        const lines = content.split('\n');
        for (let lineNum = 0; lineNum < lines.length; lineNum++) {
            const line = lines[lineNum].trim();
            
            // Skip empty lines and comments
            if (!line || line.startsWith('#')) continue;
            
            // Check for valid Show/Hide statements
            if (line.startsWith('Show') || line.startsWith('Hide')) {
                if (!this.isValidShowHide(line)) {
                    errors.push(`Line ${lineNum + 1}: Invalid Show/Hide statement`);
                }
            }
            
            // Check for valid conditions
            else if (this.isCondition(line)) {
                if (!this.isValidCondition(line)) {
                    errors.push(`Line ${lineNum + 1}: Invalid condition syntax`);
                }
            }
            
            // Check for valid actions
            else if (this.isAction(line)) {
                if (!this.isValidAction(line)) {
                    warnings.push(`Line ${lineNum + 1}: Potentially invalid action syntax`);
                }
            }
        }
        
        return {
            valid: errors.length === 0,
            errors: errors,
            warnings: warnings
        };
    }
    
    isValidShowHide(line) {
        return line === 'Show' || line === 'Hide' || 
               line.startsWith('Show ') || line.startsWith('Hide ');
    }
    
    isValidCondition(line) {
        const parts = line.split();
        return parts.length >= 2;
    }
    
    isValidAction(line) {
        const parts = line.split();
        return parts.length >= 1;
    }
    
    exportFile(data, filename) {
        const content = this.generateFilterContent(data);
        const blob = new Blob([content], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        
        const link = document.createElement('a');
        link.href = url;
        link.download = filename;
        link.style.display = 'none';
        
        document.body.appendChild(link);
        link.click();
        document.body.removeChild(link);
        
        URL.revokeObjectURL(url);
    }
    
    generateFilterContent(data) {
        let content = `# ${data.name || 'DZX Filter'}\n`;
        content += `# Generated by DZX Filter Editor\n`;
        content += `# Platform: ${data.platform || 'pc'}\n`;
        content += `# Sound Type: ${data.soundType || 'type-01'}\n\n`;
        
        const rules = data.rules || [];
        rules.forEach(rule => {
            // Add comment if exists
            if (rule.comment) {
                content += `# ${rule.comment}\n`;
            }
            
            // Add Show/Hide statement
            content += `${rule.show_hide}\n`;
            
            // Add conditions
            rule.conditions.forEach(condition => {
                const conditionLine = this.generateConditionLine(condition);
                if (conditionLine) {
                    content += `  ${conditionLine}\n`;
                }
            });
            
            // Add actions
            rule.actions.forEach(action => {
                const actionLine = this.generateActionLine(action);
                if (actionLine) {
                    content += `  ${actionLine}\n`;
                }
            });
            
            content += '\n'; // Empty line between rules
        });
        
        return content;
    }
    
    generateConditionLine(condition) {
        const conditionType = condition.type;
        const operator = condition.operator;
        const values = condition.values;
        
        if (operator === '==' && values.length === 1) {
            return `${conditionType} ${values[0]}`;
        } else if (operator === '==' && values.length > 1) {
            return `${conditionType} ${values.join(' ')}`;
        } else {
            return `${conditionType} ${operator} ${values.join(' ')}`;
        }
    }
    
    generateActionLine(action) {
        const actionType = action.type;
        const values = action.values;
        
        if (actionType.includes('Color')) {
            if (values.length >= 3) {
                let colorStr = values.slice(0, 3).join(' ');
                if (values.length > 3) {
                    colorStr += ` ${values[3]}`;
                }
                return `${actionType} ${colorStr}`;
            }
        } else if (actionType === 'CustomAlertSound') {
            if (values.length >= 2) {
                return `${actionType} "${values[0]}" ${values[1]}`;
            }
        }
        
        return `${actionType} ${values.join(' ')}`;
    }
    
    mergeFilters(filterDataList) {
        const mergedRules = [];
        let totalRules = 0;
        
        filterDataList.forEach(filterData => {
            const rules = filterData.rules || [];
            mergedRules.push(...rules);
            totalRules += rules.length;
        });
        
        return {
            rules: mergedRules,
            metadata: {
                total_rules: totalRules,
                merged_from: filterDataList.length,
                merged_at: new Date().toISOString()
            }
        };
    }
    
    // Utility methods
    hexToRgb(hex) {
        const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
        return result ? {
            r: parseInt(result[1], 16),
            g: parseInt(result[2], 16),
            b: parseInt(result[3], 16)
        } : null;
    }
    
    rgbToHex(r, g, b) {
        return "#" + ((1 << 24) + (r << 16) + (g << 8) + b).toString(16).slice(1);
    }
    
    // Advanced parsing methods
    parseAdvancedCondition(line) {
        // Handle complex conditions like "ItemLevel >= 80 && AreaLevel < 90"
        const operators = ['&&', '||', '>=', '<=', '>', '<', '==', '!='];
        const parts = line.split(/\s+/);
        
        const conditions = [];
        let currentCondition = null;
        
        for (let i = 0; i < parts.length; i++) {
            const part = parts[i];
            
            if (operators.includes(part)) {
                if (currentCondition) {
                    currentCondition.operator = part;
                }
            } else if (currentCondition && currentCondition.operator) {
                currentCondition.values.push(part);
            } else {
                if (currentCondition) {
                    conditions.push(currentCondition);
                }
                currentCondition = {
                    type: part,
                    operator: '==',
                    values: []
                };
            }
        }
        
        if (currentCondition) {
            conditions.push(currentCondition);
        }
        
        return conditions;
    }
    
    // Validation helpers
    validateRule(rule) {
        const errors = [];
        
        if (!rule.id) {
            errors.push('Rule missing ID');
        }
        
        if (!rule.show_hide || !['Show', 'Hide'].includes(rule.show_hide)) {
            errors.push('Invalid show/hide value');
        }
        
        if (!rule.conditions || !Array.isArray(rule.conditions)) {
            errors.push('Missing or invalid conditions');
        }
        
        if (!rule.actions || !Array.isArray(rule.actions)) {
            errors.push('Missing or invalid actions');
        }
        
        // Validate conditions
        rule.conditions.forEach((condition, index) => {
            if (!condition.type) {
                errors.push(`Condition ${index + 1}: missing type`);
            }
            if (!condition.operator) {
                errors.push(`Condition ${index + 1}: missing operator`);
            }
            if (!condition.values || !Array.isArray(condition.values)) {
                errors.push(`Condition ${index + 1}: missing or invalid values`);
            }
        });
        
        // Validate actions
        rule.actions.forEach((action, index) => {
            if (!action.type) {
                errors.push(`Action ${index + 1}: missing type`);
            }
            if (!action.values || !Array.isArray(action.values)) {
                errors.push(`Action ${index + 1}: missing or invalid values`);
            }
        });
        
        return {
            valid: errors.length === 0,
            errors: errors
        };
    }
    
    // Performance optimization
    optimizeRules(rules) {
        // Remove duplicate rules
        const uniqueRules = [];
        const seenRules = new Set();
        
        rules.forEach(rule => {
            const ruleKey = this.generateRuleKey(rule);
            if (!seenRules.has(ruleKey)) {
                seenRules.add(ruleKey);
                uniqueRules.push(rule);
            }
        });
        
        // Sort rules by priority (simplified)
        return uniqueRules.sort((a, b) => {
            // Rules with more conditions have higher priority
            return b.conditions.length - a.conditions.length;
        });
    }
    
    generateRuleKey(rule) {
        const conditionsKey = rule.conditions
            .map(c => `${c.type}${c.operator}${c.values.join(',')}`)
            .join('|');
        const actionsKey = rule.actions
            .map(a => `${a.type}${a.values.join(',')}`)
            .join('|');
        return `${rule.show_hide}|${conditionsKey}|${actionsKey}`;
    }
}

// Export for use in other modules
if (typeof window !== 'undefined') {
    window.BrowserFilterParser = BrowserFilterParser;
}

if (typeof module !== 'undefined' && module.exports) {
    module.exports = BrowserFilterParser;
}

File: script/filter_parser.py
# DZX Filter POE2 - Filter Parser

import re
import json
from typing import Dict, List, Any, Optional
from dataclasses import dataclass, asdict
from pathlib import Path

@dataclass
class FilterCondition:
    """Represents a condition in a filter rule"""
    type: str  # Class, BaseType, Rarity, AreaLevel, ItemLevel, etc.
    operator: str  # ==, !=, >=, <=, >, <
    values: List[str]  # List of values to match

@dataclass
class FilterAction:
    """Represents an action in a filter rule"""
    type: str  # SetTextColor, SetBorderColor, PlayEffect, etc.
    values: List[Any]  # Values for the action

@dataclass
class FilterRule:
    """Represents a complete filter rule"""
    id: str
    show_hide: str  # "Show" or "Hide"
    conditions: List[FilterCondition]
    actions: List[FilterAction]
    comment: Optional[str] = None
    enabled: bool = True

@dataclass
class FilterCategory:
    """Represents a category of filter rules"""
    name: str
    rules: List[FilterRule]
    description: Optional[str] = None

class FilterParser:
    """Parser for Path of Exile 2 filter files"""
    
    def __init__(self):
        self.conditions_pattern = re.compile(r'(\w+)\s+(.+?)(?=\s+\w+\s+|$)')
        self.actions_pattern = re.compile(r'(\w+(?:\w+\s+)*?)\s+([^\n]+)')
        self.color_pattern = re.compile(r'(\d+)\s+(\d+)\s+(\d+)(?:\s+(\d+))?')
        
    def parse_file(self, file_path: str) -> Dict[str, Any]:
        """Parse a filter file and return structured data"""
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                content = f.read()
            
            return self.parse_content(content)
        except Exception as e:
            raise Exception(f"Error parsing file {file_path}: {str(e)}")
    
    def parse_content(self, content: str) -> Dict[str, Any]:
        """Parse filter content and return structured data"""
        lines = content.split('\n')
        rules = []
        current_rule = None
        current_comment = None
        
        for line_num, line in enumerate(lines, 1):
            line = line.strip()
            
            # Skip empty lines
            if not line:
                continue
            
            # Handle comments
            if line.startswith('#'):
                current_comment = line[1:].strip()
                continue
            
            # Handle Show/Hide statements
            if line.startswith('Show') or line.startswith('Hide'):
                if current_rule:
                    rules.append(current_rule)
                
                show_hide = line.split()[0]
                current_rule = FilterRule(
                    id=f"rule_{len(rules) + 1}",
                    show_hide=show_hide,
                    conditions=[],
                    actions=[],
                    comment=current_comment
                )
                current_comment = None
                continue
            
            # Handle conditions and actions
            if current_rule:
                if self._is_condition(line):
                    condition = self._parse_condition(line)
                    if condition:
                        current_rule.conditions.append(condition)
                elif self._is_action(line):
                    action = self._parse_action(line)
                    if action:
                        current_rule.actions.append(action)
        
        # Add the last rule
        if current_rule:
            rules.append(current_rule)
        
        return {
            "rules": [asdict(rule) for rule in rules],
            "metadata": {
                "total_rules": len(rules),
                "parsed_at": str(Path().cwd())
            }
        }
    
    def _is_condition(self, line: str) -> bool:
        """Check if a line is a condition"""
        condition_keywords = [
            'Class', 'BaseType', 'Rarity', 'AreaLevel', 'ItemLevel',
            'Sockets', 'Quality', 'Width', 'Height', 'DropLevel'
        ]
        first_word = line.split()[0] if line.split() else ''
        return first_word in condition_keywords
    
    def _is_action(self, line: str) -> bool:
        """Check if a line is an action"""
        action_keywords = [
            'SetTextColor', 'SetBorderColor', 'SetBackgroundColor',
            'SetFontSize', 'PlayEffect', 'MinimapIcon', 'CustomAlertSound',
            'PlayAlertSound', 'DisableDropSound', 'Continue'
        ]
        first_word = line.split()[0] if line.split() else ''
        return first_word in action_keywords
    
    def _parse_condition(self, line: str) -> Optional[FilterCondition]:
        """Parse a condition line"""
        try:
            parts = line.split()
            if len(parts) < 2:
                return None
            
            condition_type = parts[0]
            values = []
            operator = "=="
            
            # Handle different operators
            if len(parts) >= 3:
                if parts[1] in ['>=', '<=', '>', '<', '==', '!=']:
                    operator = parts[1]
                    values = [parts[2]]
                else:
                    # Multiple values without operator
                    values = parts[1:]
            else:
                values = [parts[1]]
            
            return FilterCondition(
                type=condition_type,
                operator=operator,
                values=values
            )
        except Exception as e:
            print(f"Error parsing condition '{line}': {e}")
            return None
    
    def _parse_action(self, line: str) -> Optional[FilterAction]:
        """Parse an action line"""
        try:
            parts = line.split()
            if len(parts) < 2:
                return None
            
            action_type = parts[0]
            values = []
            
            # Handle different action types
            if action_type in ['SetTextColor', 'SetBorderColor', 'SetBackgroundColor']:
                # Parse RGB color values
                color_match = self.color_pattern.search(line)
                if color_match:
                    r, g, b = color_match.groups()[:3]
                    alpha = color_match.group(4) if color_match.group(4) else None
                    values = [int(r), int(g), int(b)]
                    if alpha:
                        values.append(int(alpha))
                else:
                    values = parts[1:]
            
            elif action_type == 'SetFontSize':
                values = [int(parts[1])]
            
            elif action_type in ['PlayEffect', 'MinimapIcon']:
                values = parts[1:]
            
            elif action_type == 'CustomAlertSound':
                # Extract sound file and volume
                sound_file = parts[1].strip('"')
                volume = int(parts[2]) if len(parts) > 2 else 100
                values = [sound_file, volume]
            
            else:
                values = parts[1:]
            
            return FilterAction(
                type=action_type,
                values=values
            )
        except Exception as e:
            print(f"Error parsing action '{line}': {e}")
            return None
    
    def validate_syntax(self, content: str) -> Dict[str, Any]:
        """Validate filter syntax and return validation results"""
        errors = []
        warnings = []
        
        lines = content.split('\n')
        for line_num, line in enumerate(lines, 1):
            line = line.strip()
            
            # Skip empty lines and comments
            if not line or line.startswith('#'):
                continue
            
            # Check for valid Show/Hide statements
            if line.startswith('Show') or line.startswith('Hide'):
                if not self._is_valid_show_hide(line):
                    errors.append(f"Line {line_num}: Invalid Show/Hide statement")
            
            # Check for valid conditions
            elif self._is_condition(line):
                if not self._is_valid_condition(line):
                    errors.append(f"Line {line_num}: Invalid condition syntax")
            
            # Check for valid actions
            elif self._is_action(line):
                if not self._is_valid_action(line):
                    warnings.append(f"Line {line_num}: Potentially invalid action syntax")
        
        return {
            "valid": len(errors) == 0,
            "errors": errors,
            "warnings": warnings
        }
    
    def _is_valid_show_hide(self, line: str) -> bool:
        """Check if Show/Hide statement is valid"""
        return line in ['Show', 'Hide'] or line.startswith('Show ') or line.startswith('Hide ')
    
    def _is_valid_condition(self, line: str) -> bool:
        """Check if condition syntax is valid"""
        parts = line.split()
        return len(parts) >= 2
    
    def _is_valid_action(self, line: str) -> bool:
        """Check if action syntax is valid"""
        parts = line.split()
        return len(parts) >= 1
    
    def export_file(self, data: Dict[str, Any], file_path: str) -> None:
        """Export structured data to filter file"""
        try:
            content = self._generate_filter_content(data)
            with open(file_path, 'w', encoding='utf-8') as f:
                f.write(content)
        except Exception as e:
            raise Exception(f"Error exporting file {file_path}: {str(e)}")
    
    def _generate_filter_content(self, data: Dict[str, Any]) -> str:
        """Generate filter file content from structured data"""
        lines = []
        
        # Add header comment
        lines.append("# DZX Filter for Path of Exile 2")
        lines.append("# Generated by DZX Filter Web Interface")
        lines.append("")
        
        rules = data.get('rules', [])
        for rule_data in rules:
            # Add comment if exists
            if rule_data.get('comment'):
                lines.append(f"# {rule_data['comment']}")
            
            # Add Show/Hide statement
            lines.append(rule_data['show_hide'])
            
            # Add conditions
            for condition in rule_data.get('conditions', []):
                condition_line = self._generate_condition_line(condition)
                if condition_line:
                    lines.append(f"  {condition_line}")
            
            # Add actions
            for action in rule_data.get('actions', []):
                action_line = self._generate_action_line(action)
                if action_line:
                    lines.append(f"  {action_line}")
            
            lines.append("")  # Empty line between rules
        
        return '\n'.join(lines)
    
    def _generate_condition_line(self, condition: Dict[str, Any]) -> str:
        """Generate condition line from condition data"""
        condition_type = condition['type']
        operator = condition['operator']
        values = condition['values']
        
        if operator == "==" and len(values) == 1:
            return f"{condition_type} {values[0]}"
        elif operator == "==" and len(values) > 1:
            return f"{condition_type} {' '.join(values)}"
        else:
            return f"{condition_type} {operator} {' '.join(values)}"
    
    def _generate_action_line(self, action: Dict[str, Any]) -> str:
        """Generate action line from action data"""
        action_type = action['type']
        values = action['values']
        
        if action_type in ['SetTextColor', 'SetBorderColor', 'SetBackgroundColor']:
            if len(values) >= 3:
                color_str = ' '.join(map(str, values[:3]))
                if len(values) > 3:
                    color_str += f" {values[3]}"
                return f"{action_type} {color_str}"
        
        elif action_type == 'CustomAlertSound':
            if len(values) >= 2:
                return f'{action_type} "{values[0]}" {values[1]}'
        
        else:
            return f"{action_type} {' '.join(map(str, values))}"
    
    def merge_filters(self, filter_data_list: List[Dict[str, Any]]) -> Dict[str, Any]:
        """Merge multiple filter data into one"""
        merged_rules = []
        total_rules = 0
        
        for filter_data in filter_data_list:
            rules = filter_data.get('rules', [])
            merged_rules.extend(rules)
            total_rules += len(rules)
        
        return {
            "rules": merged_rules,
            "metadata": {
                "total_rules": total_rules,
                "merged_from": len(filter_data_list),
                "merged_at": str(Path().cwd())
            }
        }

# Example usage
if __name__ == "__main__":
    parser = FilterParser()
    
    # Test parsing
    try:
        data = parser.parse_file("dzx_filter/filter_group/currency.filter")
        print("Parsed successfully!")
        print(f"Total rules: {data['metadata']['total_rules']}")
        
        # Test validation
        with open("dzx_filter/filter_group/currency.filter", 'r') as f:
            content = f.read()
        
        validation = parser.validate_syntax(content)
        print(f"Validation: {'Valid' if validation['valid'] else 'Invalid'}")
        if validation['errors']:
            print("Errors:", validation['errors'])
        if validation['warnings']:
            print("Warnings:", validation['warnings'])
            
    except Exception as e:
        print(f"Error: {e}")
